faas/context.go

```diff
@@ -0,0 +1,162 @@
+// Copyright 2021 Nitric Technologies Pty Ltd.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package faas
+
+import (
+	"fmt"
+
+	pb "github.com/nitrictech/apis/go/nitric/v1"
+)
+
+type TriggerContext interface {
+	Http() *HttpContext
+	Event() *EventContext
+}
+
+type triggerContextImpl struct {
+	http  *HttpContext
+	event *EventContext
+}
+
+func (t triggerContextImpl) Http() *HttpContext {
+	return t.http
+}
+
+func (t triggerContextImpl) Event() *EventContext {
+	return t.event
+}
+
+func triggerContextFromGrpcTriggerRequest(triggerReq *pb.TriggerRequest) (*triggerContextImpl, error) {
+	trigCtx := &triggerContextImpl{}
+
+	if triggerReq.GetHttp() != nil {
+		httpTrig := triggerReq.GetHttp()
+
+		headers := make(map[string][]string)
+		if httpTrig.GetHeaders() != nil {
+			for key, val := range httpTrig.GetHeaders() {
+				headers[key] = val.Value
+			}
+		} else if httpTrig.GetHeadersOld() != nil {
+			for key, val := range httpTrig.GetHeadersOld() {
+				headers[key] = []string{val}
+			}
+		}
+
+		query := make(map[string][]string)
+		if httpTrig.GetQueryParams() != nil {
+			for k, v := range httpTrig.GetQueryParams() {
+				query[k] = []string{v}
+			}
+		}
```

Comment:
Note Nitric contract query params changes may be available, once this gets through reviews.

Reply:
will be revisiting once the PR is approved in the membrane, will still need to read both pereferentially though i.e. read new if available otherwise fallback to old, so this code still applies :).